dev-tools/omdb/src/bin/omdb/mgs.rs

```diff
@@ -0,0 +1,499 @@
+#[derive(Debug, Args)]
+pub struct MgsArgs {
+    /// URL of an MGS instance to query
+    #[clap(long, env("OMDB_MGS_URL"))]
+    mgs_url: Option<String>,
+
+    #[command(subcommand)]
+    command: MgsCommands,
+}
+
+#[derive(Debug, Subcommand)]
+enum MgsCommands {
+    /// Show information about devices and components visible to MGS
+    Inventory(InventoryArgs),
+}
+
+#[derive(Debug, Args)]
+struct InventoryArgs {}
+
+impl MgsArgs {
+    pub(crate) async fn run_cmd(
+        &self,
+        omdb: &Omdb,
+        log: &slog::Logger,
+    ) -> Result<(), anyhow::Error> {
+        let mgs_url = match &self.mgs_url {
+            Some(cli_or_env_url) => cli_or_env_url.clone(),
+            None => {
+                eprintln!(
+                    "note: MGS URL not specified.  Will pick one from DNS."
+                );
+                let addrs = omdb
+                    .dns_lookup_all(
+                        log.clone(),
+                        internal_dns::ServiceName::ManagementGatewayService,
+                    )
+                    .await?;
+                let addr = addrs.into_iter().next().expect(
+                    "expected at least one MGS address from \
+                    successful DNS lookup",
+                );
+                format!("http://{}", addr)
+            }
+        };
+        eprintln!("note: using MGS URL {}", &mgs_url);
+        let mgs_client = gateway_client::Client::new(&mgs_url, log.clone());
+
+        match &self.command {
+            MgsCommands::Inventory(inventory_args) => {
+                cmd_mgs_inventory(&mgs_client, inventory_args).await
+            }
+        }
+    }
+}
+
+/// Runs `omdb mgs inventory`
+///
+/// Shows devices and components that are visible to an MGS instance.
+async fn cmd_mgs_inventory(
+    mgs_client: &gateway_client::Client,
+    _args: &InventoryArgs,
+) -> Result<(), anyhow::Error> {
+    // Report all the SP identifiers that MGS is configured to talk to.
+    println!("ALL CONFIGURED SPs\n");
+    let mut sp_ids = mgs_client
+        .sp_all_ids()
+        .await
+        .context("listing SP identifiers")?
+        .into_inner();
+    sp_ids.sort_by(sp_id_cmp);
+    show_sp_ids(&sp_ids)?;
+    println!("");
+
+    // Report which SPs are visible via Ignition.
+    println!("SPs FOUND THROUGH IGNITION\n");
+    let mut sp_list_ignition = mgs_client
+        .ignition_list()
+        .await
+        .context("listing ignition")?
+        .into_inner();
+    sp_list_ignition.sort_by(|a, b| sp_id_cmp(&a.id, &b.id));
+    show_sps_from_ignition(&sp_list_ignition)?;
+    println!("");
+
+    // Print basic state about each SP that's visible to ignition.
+    println!("SERVICE PROCESSOR STATES\n");
+    let mgs_client = std::sync::Arc::new(mgs_client);
+    let c = &mgs_client;
+    let mut sp_infos =
+        futures::stream::iter(sp_list_ignition.iter().filter_map(|ignition| {
+            if matches!(ignition.details, SpIgnition::Yes { .. }) {
```

Comment:
I'm a little unsure about using ignition as a filter on which SPs to ask for state, because at least in theory, a problem with ignition could result in ignition claiming there's nothing there but we could still talk to an SP over the management network. Similarly, ignition is fetched from MGS's local sidecar SP; if something is wrong with it, we might get no data back (at which point we'd bail out above anyway, which might or might not be reasonable).

Maybe we could run `sp_get` on all configured SPs, and ignore any errors for SPs where ignition also claims the SP is not present? This is also not super critical if you want to punt on it.

Reply:
Ahh, yeah, that does sound right. I don't think it should've taken a minute or two, but I would definitely believe a few 10s of seconds. A single missing SP should result in roughly 10 seconds to timeout based on our current config: https://github.com/oxidecomputer/omicron/blob/9aabe2aee4bbfe9af1cb9424a93cf6a59f13b9a6/smf/mgs/config.toml#L21-L26